### PR TITLE
feat: implement your test host

### PR DESCRIPTION
**Pull request by KrawczowaKris**
# Summary
## Test Task
**Implement your test host**
-  setup nginx supporting 2 servers
-  camp-php.local (this will be used for future php experiments)
-  camp-python.local (this will be used for future python experiments)
-  both servers should support named certificates for the domains mentioned above. use `mkcert cli` to do that
-  both server should implement custom location /hello, that will return `200 OK` status code along with the value of the `query_string` passed into that URL, [see here](https://www.techopedia.com/definition/1228/query-string), for example https://camp-php.local/hello?tech_stack=php, which should output the entire query_string
-  your SSL cert should be valid in CURL call in CLI, as well as if you open it in the browser
-  your camp-php.local server should output logs in JSON format
-  open PR when you done, in the PR - pls demonstrate the output of the `curl` call for the URL in each domain, and the screenshot in the browser

### camp-php.local
`curl -I https://camp-php.local`
```
HTTP/2 200 
server: nginx/1.18.0 (Ubuntu)
date: Wed, 12 Apr 2023 09:54:41 GMT
content-type: text/html
content-length: 199
last-modified: Mon, 10 Apr 2023 07:14:43 GMT
etag: "6433b763-c7"
accept-ranges: bytes
```

screenshot in the browser:
изображение

logs in json format:
```
{ "@timestamp": "2023-04-12T14:54:41+05:00", "remote_addr": "127.0.0.1", "body_bytes_sent": "0", "status": 200, "request": "HEAD / HTTP/2.0", "url": "/index.html", "request_method": "HEAD", "response_time": 0, "http_referrer": "", "http_user_agent": "curl/7.81.0" }
{ "@timestamp": "2023-04-12T14:57:06+05:00", "remote_addr": "127.0.0.1", "body_bytes_sent": "14", "status": 200, "request": "GET /hello?tech_stack=php HTTP/2.0", "url": "/hello", "request_method": "GET", "response_time": 0, "http_referrer": "", "http_user_agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36" }
```

### camp-python.local
`curl -I https://camp-python.local`
```
HTTP/2 200 
server: nginx/1.18.0 (Ubuntu)
date: Wed, 12 Apr 2023 10:01:08 GMT
content-type: text/html
content-length: 202
last-modified: Mon, 10 Apr 2023 07:14:12 GMT
etag: "6433b744-ca"
accept-ranges: bytes
```

screenshot in the browser:
изображение

### CODE REVIEW

File: chapter2/tests/curl-camp-php.log
Comment (dmitry-mightydevops) [inline chapter2/tests/curl-camp-php.log:1]: no need to store artifacts such as logs. You can attach them in the PR - but don't store in the remote. They're noise.

File: chapter2/tests/curl-camp-python.log
Comment (dmitry-mightydevops) [inline chapter2/tests/curl-camp-python.log:1]: Noise in the repo

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]
I only see the log files in the PR - no actual implementation

File: chapter2/tests/curl-camp-php.log
Comment (KrawczowaKris) [reply chapter2/tests/curl-camp-php.log:1]: done

File: chapter2/tests/camp-php.local [hunk 1]
```diff
@@ -0,0 +1,40 @@
+map $upstream_response_time $temprt {
```
[the remark this answers is not in the document]
Comment (KrawczowaKris) [reply chapter2/tests/camp-php.local]: remade

File: chapter2/tests/nginx.conf [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		location / {
+			try_files $uri $uri/ =404;
+		}
+
+		location = /hello {
+			add_header Content-Type text/plain;
+ 			return 200 $args;
+		}
+
+		access_log /var/log/nginx/camp-php.local_access.log json;
```
Comment (dmitry-mightydevops) [inline chapter2/tests/nginx.conf:54]: prefer nothing after you declare `location`. I.e. all your logs should be right after the listen.

File: chapter2/tests/nginx.conf [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+				'"body_bytes_sent": "$body_bytes_sent", '
+				'"status": "$status", '
+				'"request": "$request", '
+				'"url": "$uri", '
+				'"request_method": "$request_method", '
+				'"http_user_agent": "$http_user_agent" }';
+
+	server {
+		listen 80;
+		listen [::]:80;
```
Comment (dmitry-mightydevops) [inline chapter2/tests/nginx.conf:34]: no need for 80 if you have a superior 443.

File: chapter2/tests/nginx.conf [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+		listen [::]:443 ssl http2;
+		ssl_certificate /tmp/camp-python.local.pem;
+		ssl_certificate_key /tmp/camp-python.local-key.pem;
+
+		root /var/www/camp-python.local/html;
+		index index.html index.htm index.nginx-debian.html;
+		server_name camp-python.local www.camp-python.local;
+
+		location / {
+		try_files $uri $uri/ =404;
```
Comment (dmitry-mightydevops) [inline chapter2/tests/nginx.conf:71]: formatting issue.

File: chapter2/tests/nginx.conf [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+http {
+	sendfile on;
+	tcp_nopush on;
+	types_hash_max_size 4096;
+	include /etc/nginx/mime.types;
+	default_type application/octet-stream;
+	include /etc/nginx/conf.d/*.conf;
+
+	log_format json '{ "@timestamp": "$time_iso8601", '
```
Comment (dmitry-mightydevops) [inline chapter2/tests/nginx.conf:23]: make nicer format

```
log_format json_combined escape=json
  '{'
    '"time_local":"$time_local",'
    '"remote_addr":"$remote_addr",'
    '"remote_user":"$remote_user",'
    '"request":"$request",'
    '"status": "$status",'
    '"body_bytes_sent":"$body_bytes_sent",'
    '"request_time":"$request_time",'
    '"http_referrer":"$http_referer",'
    '"http_user_agent":"$http_user_agent"'
  '}';
```

File: chapter2/tests/nginx.conf [hunk 6]
```diff
@@ -0,0 +1,79 @@
+user nginx nginx;
+
+worker_processes auto; 
+
+error_log /var/log/nginx/error.log notice;
+error_log /var/log/nginx/error_debug.log debug;  
+
+pid /run/nginx.pid;
+include /usr/share/nginx/modules/*.conf;
```
Comment (dmitry-mightydevops) [inline chapter2/tests/nginx.conf:9]: no need for includes, pls always supply self-sufficient single nginx.conf for the sake of debugging.

File: chapter2/tests/nginx.conf [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	worker_connections 1024;
+}
+
+http {
+	sendfile on;
+	tcp_nopush on;
+	types_hash_max_size 4096;
+	include /etc/nginx/mime.types;
+	default_type application/octet-stream;
+	include /etc/nginx/conf.d/*.conf;
```
Comment (dmitry-mightydevops) [inline chapter2/tests/nginx.conf:21]: no need for includes for test tasks

File: chapter2/tests/nginx.conf [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+events {
+	worker_connections 1024;
+}
+
+http {
+	sendfile on;
+	tcp_nopush on;
+	types_hash_max_size 4096;
+	include /etc/nginx/mime.types;
```
Comment (dmitry-mightydevops) [inline chapter2/tests/nginx.conf:19]: that's fine to keep it *the only include you can keep it.

File: chapter2/tests/nginx.conf [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+	default_type application/octet-stream;
```
Comment (dmitry-mightydevops) [inline chapter2/tests/nginx.conf:20]: if you have multiple "similar" directives - group them.

i.e.

```
include....
default_type
include...
```

is not good for reading.

Same issue with logs. You define the `log_format` and `access_log` in different areas of your `server` context. They should be close to each other.

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: chapter2/tests/nginx.conf (hunk 2 above)
Comment (KrawczowaKris) [reply chapter2/tests/nginx.conf:54]: done

File: chapter2/tests/nginx.conf (hunk 3 above)
Comment (KrawczowaKris) [reply chapter2/tests/nginx.conf:34]: done

File: chapter2/tests/nginx.conf (hunk 4 above)
Comment (KrawczowaKris) [reply chapter2/tests/nginx.conf:71]: done

File: chapter2/tests/nginx.conf (hunk 6 above)
Comment (KrawczowaKris) [reply chapter2/tests/nginx.conf:9]: removed

File: chapter2/tests/nginx.conf (hunk 7 above)
Comment (KrawczowaKris) [reply chapter2/tests/nginx.conf:21]: removed

File: chapter2/tests/nginx.conf (hunk 9 above)
Comment (KrawczowaKris) [reply chapter2/tests/nginx.conf:20]: remade

File: chapter2/tests/nginx.conf (hunk 5 above)
Comment (KrawczowaKris) [reply chapter2/tests/nginx.conf:23]: better now?

**Review by dmitry-mightydevops — APPROVED** [review]

File: chapter2/tests/nginx.conf (hunk 5 above)
Comment (dmitry-mightydevops) [reply chapter2/tests/nginx.conf]: yes much :)

**Pull request merged**